src/Lykke.AlgoStore.CSharp.AlgoTemplate.Services/Functions/ADX/AdxFunction.cs

```diff
@@ -0,0 +1,133 @@
+﻿using Lykke.AlgoStore.CSharp.Algo.Core.Candles;
+using Lykke.AlgoStore.CSharp.Algo.Core.Functions;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Lykke.AlgoStore.CSharp.AlgoTemplate.Services.Functions.ADX
+{
+    public class AdxFunction : IFunction
+    {
+        private readonly int _period;
+        private int _samples;
+        private Candle _previousInput;
+        private AdxParameters _functionParams = new AdxParameters();
+        public FunctionParamsBase FunctionParameters => _functionParams;
+
+        #region Additional Parameters
+
+        private double? DirectionalMovementIndexMinus { get; set; }
+        private double? DirectionalMovementIndexPlus { get; set; }
+
+        private double DirectionalMovementIndex { get; set; }
+        private Queue<double> DirectionalMovementIndexes { get; set; }
+
+        private double? PreviousADX { get; set; }
+        public double? AverageTrueRange { get; set; }
+
+        private DirectionalMovementIndexPlusFunction DMIPlusFucn { get; set; }
+        private DirectionalMovementIndexMinusFunction DMIMinusFucn { get; set; }
+        private ATRFunction ATRFunction { get; set; }
+        #endregion
+
+        public AdxFunction(AdxParameters adxParameters)
+        {
+            _period = adxParameters.AdxPeriod;
+            _functionParams = adxParameters;
+
+            DirectionalMovementIndexes = _functionParams.AdxPeriod == 0 ? new Queue<double>() : new Queue<double>(_functionParams.AdxPeriod);
+            AverageTrueRange = 0.0d;
+
+            DMIPlusFucn = new DirectionalMovementIndexPlusFunction(new DMIParameters() { Period = _period, IsAverageTrueRangeSet = true });
+            DMIMinusFucn = new DirectionalMovementIndexMinusFunction(new DMIParameters() { Period = _period, IsAverageTrueRangeSet = true });
+            ATRFunction = new ATRFunction(new AtrParameters() { Period = _period });
+        }
+
+        public double? WarmUp(IList<Candle> values)
+        {
+            if (values == null)
+                throw new ArgumentException();
+
+            foreach (var value in values)
+            {
+                _samples++;
+
+                AverageTrueRange = ATRFunction.AddNewValue(value);
+
+                DMIPlusFucn.AverageTrueRange = AverageTrueRange;
+                DMIMinusFucn.AverageTrueRange = AverageTrueRange;
+
+                DirectionalMovementIndexPlus = DMIPlusFucn.AddNewValue(value);
+                DirectionalMovementIndexMinus = DMIMinusFucn.AddNewValue(value);
+
+                if (_samples >= _period + 1)
+                {
+                    DirectionalMovementIndex = (Math.Abs(DirectionalMovementIndexPlus.Value - DirectionalMovementIndexMinus.Value)
```

Comment:
Remove parantheses around the first Math.Max.  Is there a possibility for DirectionalMovementIndexPlus and DirectionalMovementIndexMinus to be null. Maybe check for null.

Reply:
Remove parentheses around the first Math.Max - parentheses are used because I need to multiply the whole result. I think otherwise multiplay will be executed before dividing...

Reply:
Value probably can't be null because DirectionalMovementIndex  is calculated after several calculations of Directional movement....DirectionalMovementIndexPlus will be null only the first time, but here in this case we enter when period *2

![image](https://user-images.githubusercontent.com/31039385/36673511-a8332e72-1b0b-11e8-9ab7-62eb3b7d9644.png)